In [2]:
import pandas as pd
from pathlib import Path

In [6]:
from pathlib import Path

PROJECT_ROOT = Path.cwd().parent
RAW_DATA_DIR = PROJECT_ROOT / "data" / "raw"

print("Notebook location:", Path.cwd())
print("Raw data location:", RAW_DATA_DIR)
print("\nFiles found:")

for file in RAW_DATA_DIR.iterdir():
    print(file.name)

Notebook location: d:\Projects\SW2627-PandasNumPy-OnboardIQ\notebooks
Raw data location: d:\Projects\SW2627-PandasNumPy-OnboardIQ\data\raw

Files found:
.gitkeep
employee.csv
internal_tool_usage.csv
learning_management.csv
onboarding_progress.csv
support_tickets.csv


In [7]:
datasets = {
    "Employee": pd.read_csv(RAW_DATA_DIR / "employee.csv"),
    "Onboarding Progress": pd.read_csv(
        RAW_DATA_DIR / "onboarding_progress.csv"
    ),
    "Internal Tool Usage": pd.read_csv(
        RAW_DATA_DIR / "internal_tool_usage.csv"
    ),
    "Support Tickets": pd.read_csv(
        RAW_DATA_DIR / "support_tickets.csv"
    ),
    "Learning Management": pd.read_csv(
        RAW_DATA_DIR / "learning_management.csv"
    )
}

In [8]:
for name, df in datasets.items():
    print(name, "->", df.shape)

Employee -> (311, 36)
Onboarding Progress -> (4222, 9)
Internal Tool Usage -> (13995, 9)
Support Tickets -> (1048, 8)
Learning Management -> (3192, 9)


In [9]:
for name, df in datasets.items():
    print(f"\n{'=' * 60}")
    print(name)
    print(f"{'=' * 60}")
    print("Columns:")
    for column in df.columns:
        print(f"  - {column}")


Employee
Columns:
  - Employee_Name
  - EmpID
  - MarriedID
  - MaritalStatusID
  - GenderID
  - EmpStatusID
  - DeptID
  - PerfScoreID
  - FromDiversityJobFairID
  - Salary
  - Termd
  - PositionID
  - Position
  - State
  - Zip
  - DOB
  - Sex
  - MaritalDesc
  - CitizenDesc
  - HispanicLatino
  - RaceDesc
  - DateofHire
  - DateofTermination
  - TermReason
  - EmploymentStatus
  - Department
  - ManagerName
  - ManagerID
  - RecruitmentSource
  - PerformanceScore
  - EngagementSurvey
  - EmpSatisfaction
  - SpecialProjectsCount
  - LastPerformanceReview_Date
  - DaysLateLast30
  - Absences

Onboarding Progress
Columns:
  - onboarding_task_id
  - employee_id
  - task_name
  - task_category
  - current_stage
  - task_status
  - due_date
  - completed_date
  - completion_percentage

Internal Tool Usage
Columns:
  - usage_id
  - employee_id
  - usage_date
  - slack_activity_count
  - jira_usage_count
  - github_activity_count
  - teams_activity_count
  - workspace_activity_count
  - da

In [10]:
print("EMPLOYEE DATASET")
print("Number of columns:", len(datasets["Employee"].columns))

for i, column in enumerate(datasets["Employee"].columns, start=1):
    print(f"{i}. {column}")

EMPLOYEE DATASET
Number of columns: 36
1. Employee_Name
2. EmpID
3. MarriedID
4. MaritalStatusID
5. GenderID
6. EmpStatusID
7. DeptID
8. PerfScoreID
9. FromDiversityJobFairID
10. Salary
11. Termd
12. PositionID
13. Position
14. State
15. Zip
16. DOB
17. Sex
18. MaritalDesc
19. CitizenDesc
20. HispanicLatino
21. RaceDesc
22. DateofHire
23. DateofTermination
24. TermReason
25. EmploymentStatus
26. Department
27. ManagerName
28. ManagerID
29. RecruitmentSource
30. PerformanceScore
31. EngagementSurvey
32. EmpSatisfaction
33. SpecialProjectsCount
34. LastPerformanceReview_Date
35. DaysLateLast30
36. Absences


In [11]:
print("ONBOARDING PROGRESS DATASET")
print("Number of columns:", len(datasets["Onboarding Progress"].columns))

for i, column in enumerate(datasets["Onboarding Progress"].columns, start=1):
    print(f"{i}. {column}")

ONBOARDING PROGRESS DATASET
Number of columns: 9
1. onboarding_task_id
2. employee_id
3. task_name
4. task_category
5. current_stage
6. task_status
7. due_date
8. completed_date
9. completion_percentage


In [12]:
print("INTERNAL TOOL USAGE DATASET")
print("Number of columns:", len(datasets["Internal Tool Usage"].columns))

for i, column in enumerate(datasets["Internal Tool Usage"].columns, start=1):
    print(f"{i}. {column}")

INTERNAL TOOL USAGE DATASET
Number of columns: 9
1. usage_id
2. employee_id
3. usage_date
4. slack_activity_count
5. jira_usage_count
6. github_activity_count
7. teams_activity_count
8. workspace_activity_count
9. daily_active_usage


In [13]:
print("SUPPORT TICKET DATASET")
print("Number of columns:", len(datasets["Support Tickets"].columns))

for i, column in enumerate(datasets["Support Tickets"].columns, start=1):
    print(f"{i}. {column}")

SUPPORT TICKET DATASET
Number of columns: 8
1. ticket_id
2. employee_id
3. created_date
4. issue_category
5. priority
6. resolution_time_hours
7. assigned_team
8. status


In [14]:
print("LEARNING MANAGEMENT DATASET")
print("Number of columns:", len(datasets["Learning Management"].columns))

for i, column in enumerate(datasets["Learning Management"].columns, start=1):
    print(f"{i}. {column}")

LEARNING MANAGEMENT DATASET
Number of columns: 9
1. learning_record_id
2. employee_id
3. course_name
4. course_type
5. assigned_date
6. completed_date
7. assessment_score
8. completion_time_hours
9. course_status


In [16]:
employee = datasets["Employee"]
onboarding = datasets["Onboarding Progress"]
tool_usage = datasets["Internal Tool Usage"]
support = datasets["Support Tickets"]
learning = datasets["Learning Management"]

In [17]:
print(employee["EmpID"].head())
print(employee["EmpID"].dtype)

print("\nOnboarding employee IDs:")
print(onboarding["employee_id"].head())
print(onboarding["employee_id"].dtype)

0    10026
1    10084
2    10196
3    10088
4    10069
Name: EmpID, dtype: int64
int64

Onboarding employee IDs:
0    10026
1    10026
2    10026
3    10026
4    10026
Name: employee_id, dtype: int64
int64


In [18]:
employee_ids = set(employee["EmpID"].dropna())
onboarding_ids = set(onboarding["employee_id"].dropna())
tool_usage_ids = set(tool_usage["employee_id"].dropna())
support_ids = set(support["employee_id"].dropna())
learning_ids = set(learning["employee_id"].dropna())

print("Onboarding IDs not in Employee:",
      len(onboarding_ids - employee_ids))

print("Tool Usage IDs not in Employee:",
      len(tool_usage_ids - employee_ids))

print("Support Ticket IDs not in Employee:",
      len(support_ids - employee_ids))

print("Learning IDs not in Employee:",
      len(learning_ids - employee_ids))

Onboarding IDs not in Employee: 0
Tool Usage IDs not in Employee: 0
Support Ticket IDs not in Employee: 0
Learning IDs not in Employee: 0


## Referential Integrity Check

Employee ID was validated as the primary integration key across all onboarding-related datasets.

Results:

| Dataset | Orphan Employee IDs | Result |
|---|---:|---|
| Onboarding Progress | 0 | Pass |
| Internal Tool Usage | 0 | Pass |
| Support Tickets | 0 | Pass |
| Learning Management | 0 | Pass |

All employee IDs in the dependent datasets are present in the Employee master dataset. No orphan records were identified during the initial profiling stage.


In [19]:
unique_keys = {
    "Employee": ("EmpID", employee),
    "Onboarding Progress": ("onboarding_task_id", onboarding),
    "Internal Tool Usage": ("usage_id", tool_usage),
    "Support Tickets": ("ticket_id", support),
    "Learning Management": ("learning_record_id", learning)
}

for name, (key, df) in unique_keys.items():
    duplicate_count = df[key].duplicated().sum()
    null_count = df[key].isna().sum()

    print(f"\n{name}")
    print(f"  Key: {key}")
    print(f"  Duplicate IDs: {duplicate_count}")
    print(f"  Null IDs: {null_count}")


Employee
  Key: EmpID
  Duplicate IDs: 0
  Null IDs: 0

Onboarding Progress
  Key: onboarding_task_id
  Duplicate IDs: 0
  Null IDs: 0

Internal Tool Usage
  Key: usage_id
  Duplicate IDs: 0
  Null IDs: 0

Support Tickets
  Key: ticket_id
  Duplicate IDs: 0
  Null IDs: 0

Learning Management
  Key: learning_record_id
  Duplicate IDs: 0
  Null IDs: 0


## Primary Key Validation

All datasets were checked for duplicate and null values in their designated primary key fields.

| Dataset | Primary Key | Duplicate IDs | Null IDs | Result |
|---|---|---:|---:|---|
| Employee | `EmpID` | 0 | 0 | Pass |
| Onboarding Progress | `onboarding_task_id` | 0 | 0 | Pass |
| Internal Tool Usage | `usage_id` | 0 | 0 | Pass |
| Support Tickets | `ticket_id` | 0 | 0 | Pass |
| Learning Management | `learning_record_id` | 0 | 0 | Pass |

No duplicate or null primary-key values were identified during initial profiling.

In [20]:
for name, df in datasets.items():
    print(f"\n{'=' * 60}")
    print(name)
    print(f"{'=' * 60}")

    missing = df.isna().sum()
    missing_pct = (missing / len(df) * 100).round(2)

    result = pd.DataFrame({
        "Missing Values": missing,
        "Missing %": missing_pct
    })

    print(result[result["Missing Values"] > 0])


Employee
                   Missing Values  Missing %
DateofTermination             207      66.56
ManagerID                       8       2.57

Onboarding Progress
                Missing Values  Missing %
completed_date            1142      27.05

Internal Tool Usage
Empty DataFrame
Columns: [Missing Values, Missing %]
Index: []

Support Tickets
                       Missing Values  Missing %
resolution_time_hours             105      10.02

Learning Management
                       Missing Values  Missing %
completed_date                    961      30.11
assessment_score                  840      26.32
completion_time_hours             961      30.11


In [21]:
for name, df in datasets.items():
    print(f"\n{'=' * 70}")
    print(name)
    print(f"{'=' * 70}")

    missing = df.isna().sum()
    missing_pct = (missing / len(df) * 100).round(2)

    result = pd.DataFrame({
        "Missing Values": missing,
        "Missing %": missing_pct
    })

    result = result[result["Missing Values"] > 0]

    if result.empty:
        print("No missing values.")
    else:
        print(result.to_string())


Employee
                   Missing Values  Missing %
DateofTermination             207      66.56
ManagerID                       8       2.57

Onboarding Progress
                Missing Values  Missing %
completed_date            1142      27.05

Internal Tool Usage
No missing values.

Support Tickets
                       Missing Values  Missing %
resolution_time_hours             105      10.02

Learning Management
                       Missing Values  Missing %
completed_date                    961      30.11
assessment_score                  840      26.32
completion_time_hours             961      30.11


In [22]:
# Support ticket check
print("SUPPORT TICKETS")
print(
    support.groupby("status")["resolution_time_hours"]
    .apply(lambda x: x.isna().sum())
)

print("\nLearning Management")
print(
    learning.groupby("course_status")[[
        "completed_date",
        "assessment_score",
        "completion_time_hours"
    ]].apply(lambda x: x.isna().sum())
)

print("\nOnboarding Progress")
print(
    onboarding.groupby("task_status")["completed_date"]
    .apply(lambda x: x.isna().sum())
)

SUPPORT TICKETS
status
Closed          0
In Progress    83
Open           22
Resolved        0
Name: resolution_time_hours, dtype: int64

Learning Management
               completed_date  assessment_score  completion_time_hours
course_status                                                         
Assigned                  288               288                    288
Completed                   0                 0                      0
In Progress               447               326                    447
Overdue                   226               226                    226

Onboarding Progress
task_status
Blocked        282
Complete         0
In Progress    504
Not Started    154
Overdue        202
Name: completed_date, dtype: int64


## Business-Rule Validation of Missing Values

Missing values were evaluated against workflow status to distinguish expected nulls from potential data-quality issues.

### Support Tickets

`resolution_time_hours` is missing only for tickets with `Open` or `In Progress` status. All `Resolved` and `Closed` tickets have a resolution time.

**Result: Pass.** Missing resolution times are consistent with unresolved ticket states.

### Learning Management

`completed_date` and `completion_time_hours` are missing for `Assigned`, `In Progress`, and `Overdue` records, while all `Completed` records contain both fields.

**Result: Pass.** Missing completion information is consistent with course status.

`assessment_score` contains some null values for records that have not completed or may not yet require assessment.

**Result: Review during processing.** Assessment scores should remain null when an assessment is not applicable or has not yet been completed.

### Onboarding Progress

`completed_date` is missing for `Blocked`, `In Progress`, `Not Started`, and `Overdue` tasks. All `Complete` tasks have a completion date.

**Result: Pass.** Missing completion dates are consistent with incomplete task statuses.

Overall, the identified missing values are primarily workflow-dependent nulls rather than invalid records. They should not be blindly imputed or removed during data cleaning.

In [23]:
for name, df in datasets.items():
    print(f"\n{'=' * 70}")
    print(name)
    print(f"{'=' * 70}")
    print(df.dtypes.to_string())


Employee
Employee_Name                     str
EmpID                           int64
MarriedID                       int64
MaritalStatusID                 int64
GenderID                        int64
EmpStatusID                     int64
DeptID                          int64
PerfScoreID                     int64
FromDiversityJobFairID          int64
Salary                          int64
Termd                           int64
PositionID                      int64
Position                          str
State                             str
Zip                             int64
DOB                               str
Sex                               str
MaritalDesc                       str
CitizenDesc                       str
HispanicLatino                    str
RaceDesc                          str
DateofHire                        str
DateofTermination                 str
TermReason                        str
EmploymentStatus                  str
Department                        str
Ma

In [24]:
date_columns = {
    "Employee": ["DOB", "DateofHire", "DateofTermination"],
    "Onboarding Progress": ["due_date", "completed_date"],
    "Internal Tool Usage": ["usage_date"],
    "Support Tickets": ["created_date"],
    "Learning Management": ["assigned_date", "completed_date"]
}

for name, columns in date_columns.items():
    df = datasets[name]

    print(f"\n{'=' * 60}")
    print(name)
    print(f"{'=' * 60}")

    for column in columns:
        print(f"{column}: {df[column].dtype}")


Employee
DOB: str
DateofHire: str
DateofTermination: str

Onboarding Progress
due_date: str
completed_date: str

Internal Tool Usage
usage_date: str

Support Tickets
created_date: str

Learning Management
assigned_date: str
completed_date: str


In [25]:
for name, columns in date_columns.items():
    df = datasets[name]

    print(f"\n{'=' * 60}")
    print(name)
    print(f"{'=' * 60}")

    for column in columns:
        parsed = pd.to_datetime(df[column], errors="coerce")

        invalid = parsed.notna().sum()
        failed = df[column].notna().sum() - invalid

        print(
            f"{column}: "
            f"valid={invalid}, "
            f"invalid={failed}, "
            f"null={df[column].isna().sum()}"
        )


Employee
DOB: valid=311, invalid=0, null=0
DateofHire: valid=311, invalid=0, null=0
DateofTermination: valid=104, invalid=0, null=207

Onboarding Progress
due_date: valid=4222, invalid=0, null=0
completed_date: valid=3080, invalid=0, null=1142

Internal Tool Usage
usage_date: valid=13995, invalid=0, null=0

Support Tickets
created_date: valid=1048, invalid=0, null=0

Learning Management
assigned_date: valid=3192, invalid=0, null=0
completed_date: valid=2231, invalid=0, null=961


C:\Users\Sapna\AppData\Local\Temp\ipykernel_18780\1846641369.py:9: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  parsed = pd.to_datetime(df[column], errors="coerce")


## Date Field Validation

All date and timestamp fields were tested for parseability using Pandas datetime conversion with invalid values coerced for validation purposes.

| Dataset | Field | Valid | Invalid | Null | Result |
|---|---|---:|---:|---:|---|
| Employee | `DOB` | 311 | 0 | 0 | Pass |
| Employee | `DateofHire` | 311 | 0 | 0 | Pass |
| Employee | `DateofTermination` | 104 | 0 | 207 | Pass |
| Onboarding Progress | `due_date` | 4,222 | 0 | 0 | Pass |
| Onboarding Progress | `completed_date` | 3,080 | 0 | 1,142 | Pass |
| Internal Tool Usage | `usage_date` | 13,995 | 0 | 0 | Pass |
| Support Tickets | `created_date` | 1,048 | 0 | 0 | Pass |
| Learning Management | `assigned_date` | 3,192 | 0 | 0 | Pass |
| Learning Management | `completed_date` | 2,231 | 0 | 961 | Pass |

No invalid populated date values were identified. Null dates were primarily associated with records that have not yet reached a completion or termination state.

The raw CSV date fields are currently stored as strings. They will be converted to appropriate datetime types during the processing and transformation stage rather than modifying the raw source files.

In [26]:
categorical_columns = {
    "Employee": [
        "Department",
        "Position",
        "State",
        "Sex",
        "EmpStatusID"
    ],
    "Onboarding Progress": [
        "task_category",
        "current_stage",
        "task_status"
    ],
    "Internal Tool Usage": [],
    "Support Tickets": [
        "issue_category",
        "priority",
        "assigned_team",
        "status"
    ],
    "Learning Management": [
        "course_type",
        "course_status"
    ]
}

for name, columns in categorical_columns.items():
    df = datasets[name]

    print(f"\n{'=' * 70}")
    print(name)
    print(f"{'=' * 70}")

    for column in columns:
        print(f"\n{column}:")
        print(df[column].value_counts(dropna=False).to_string())


Employee

Department:
Department
Production              209
IT/IS                    50
Sales                    31
Software Engineering     11
Admin Offices             9
Executive Office          1

Position:
Position
Production Technician I         137
Production Technician II         57
Area Sales Manager               27
Production Manager               14
Software Engineer                10
IT Support                        8
Data Analyst                      7
Database Administrator            5
Sr. Network Engineer              5
Network Engineer                  5
BI Developer                      4
Accountant I                      3
Sales Manager                     3
Administrative Assistant          3
Senior BI Developer               3
Sr. DBA                           2
Sr. Accountant                    2
IT Manager - DB                   2
Enterprise Architect              1
Software Engineering Manager      1
BI Director                       1
Director of Operations

In [27]:
for name, columns in categorical_columns.items():
    df = datasets[name]

    print(f"\n{'=' * 70}")
    print(name)
    print(f"{'=' * 70}")

    for column in columns:
        values = sorted(df[column].dropna().astype(str).unique())

        print(f"\n{column} ({len(values)} unique values):")
        print(values)
        


Employee

Department (6 unique values):
['Admin Offices', 'Executive Office', 'IT/IS', 'Production       ', 'Sales', 'Software Engineering']

Position (32 unique values):
['Accountant I', 'Administrative Assistant', 'Area Sales Manager', 'BI Developer', 'BI Director', 'CIO', 'Data Analyst', 'Data Analyst ', 'Data Architect', 'Database Administrator', 'Director of Operations', 'Director of Sales', 'Enterprise Architect', 'IT Director', 'IT Manager - DB', 'IT Manager - Infra', 'IT Manager - Support', 'IT Support', 'Network Engineer', 'President & CEO', 'Principal Data Architect', 'Production Manager', 'Production Technician I', 'Production Technician II', 'Sales Manager', 'Senior BI Developer', 'Shared Services Manager', 'Software Engineer', 'Software Engineering Manager', 'Sr. Accountant', 'Sr. DBA', 'Sr. Network Engineer']

State (28 unique values):
['AL', 'AZ', 'CA', 'CO', 'CT', 'FL', 'GA', 'ID', 'IN', 'KY', 'MA', 'ME', 'MT', 'NC', 'ND', 'NH', 'NV', 'NY', 'OH', 'OR', 'PA', 'RI', 'TN'

In [28]:
for name, columns in {
    "Onboarding Progress": [
        "task_category",
        "current_stage",
        "task_status"
    ],
    "Support Tickets": [
        "issue_category",
        "priority",
        "status"
    ],
    "Learning Management": [
        "course_type",
        "course_status"
    ]
}.items():

    df = datasets[name]

    print(f"\n{'=' * 60}")
    print(name)
    print(f"{'=' * 60}")

    for column in columns:
        print(f"\n{column}:")
        print(df[column].value_counts(dropna=False).to_string())


Onboarding Progress

task_category:
task_category
IT               1202
HR                911
Manager           908
Role-specific     601
Compliance        600

current_stage:
current_stage
Enablement     1810
Setup           899
Preboarding     607
Orientation     606
Productive      300

task_status:
task_status
Complete       3080
In Progress     504
Blocked         282
Overdue         202
Not Started     154

Support Tickets

issue_category:
issue_category
Access       313
Hardware     219
Learning     152
Other        136
HR policy    119
Payroll      109

priority:
priority
Medium      565
Low         256
High        186
Critical     41

status:
status
Resolved       727
Closed         216
In Progress     83
Open            22

Learning Management

course_type:
course_type
Compliance       1201
Orientation       933
Role-specific     933
Optional          125

course_status:
course_status
Completed      2231
In Progress     447
Assigned        288
Overdue         226


In [29]:
numeric_rules = {
    "Onboarding Progress": {
        "completion_percentage": (0, 100)
    },
    "Internal Tool Usage": {
        "slack_activity_count": (0, None),
        "jira_usage_count": (0, None),
        "github_activity_count": (0, None),
        "teams_activity_count": (0, None),
        "workspace_activity_count": (0, None)
    },
    "Support Tickets": {
        "resolution_time_hours": (0, None)
    },
    "Learning Management": {
        "assessment_score": (0, 100),
        "completion_time_hours": (0, None)
    }
}

for dataset_name, columns in numeric_rules.items():
    df = datasets[dataset_name]

    print(f"\n{'=' * 60}")
    print(dataset_name)
    print(f"{'=' * 60}")

    for column, (minimum, maximum) in columns.items():
        values = pd.to_numeric(df[column], errors="coerce")

        below_min = (values < minimum).sum()

        if maximum is not None:
            above_max = (values > maximum).sum()
        else:
            above_max = 0

        print(
            f"{column}: "
            f"below_min={below_min}, "
            f"above_max={above_max}, "
            f"min={values.min()}, "
            f"max={values.max()}"
        )


Onboarding Progress
completion_percentage: below_min=0, above_max=0, min=0.0, max=100.0

Internal Tool Usage
slack_activity_count: below_min=0, above_max=0, min=0, max=19
jira_usage_count: below_min=0, above_max=0, min=0, max=12
github_activity_count: below_min=0, above_max=0, min=0, max=14
teams_activity_count: below_min=0, above_max=0, min=0, max=13
workspace_activity_count: below_min=0, above_max=0, min=0, max=21

Support Tickets
resolution_time_hours: below_min=0, above_max=0, min=1.05, max=112.88

Learning Management
assessment_score: below_min=0, above_max=0, min=40.0, max=100.0
completion_time_hours: below_min=0, above_max=0, min=0.86, max=9.09


## Numeric Range Validation

Numeric fields were validated against their expected business ranges.

| Dataset | Field | Minimum | Maximum | Range Violations | Result |
|---|---|---:|---:|---:|---|
| Onboarding Progress | `completion_percentage` | 0 | 100 | 0 | Pass |
| Internal Tool Usage | `slack_activity_count` | 0 | 19 | 0 | Pass |
| Internal Tool Usage | `jira_usage_count` | 0 | 12 | 0 | Pass |
| Internal Tool Usage | `github_activity_count` | 0 | 14 | 0 | Pass |
| Internal Tool Usage | `teams_activity_count` | 0 | 13 | 0 | Pass |
| Internal Tool Usage | `workspace_activity_count` | 0 | 21 | 0 | Pass |
| Support Tickets | `resolution_time_hours` | 1.05 | 112.88 | 0 | Pass |
| Learning Management | `assessment_score` | 40 | 100 | 0 | Pass |
| Learning Management | `completion_time_hours` | 0.86 | 9.09 | 0 | Pass |

No numeric range violations were identified. All activity counts and duration fields are non-negative, while percentage and assessment fields remain within their defined 0–100 ranges.

In [30]:
employee_numeric_rules = {
    "Salary": (0, None),
    "Absences": (0, None),
    "DaysLateLast30": (0, None),
    "SpecialProjectsCount": (0, None),
    "PerfScoreID": (1, None),
    "EmpStatusID": (1, None)
}

employee = datasets["Employee"]

print(f"{'=' * 60}")
print("Employee Numeric Validation")
print(f"{'=' * 60}")

for column, (minimum, maximum) in employee_numeric_rules.items():
    values = pd.to_numeric(employee[column], errors="coerce")

    below_min = (values < minimum).sum()

    if maximum is not None:
        above_max = (values > maximum).sum()
    else:
        above_max = 0

    print(
        f"{column}: "
        f"below_min={below_min}, "
        f"above_max={above_max}, "
        f"min={values.min()}, "
        f"max={values.max()}"
    )

Employee Numeric Validation
Salary: below_min=0, above_max=0, min=45046, max=250000
Absences: below_min=0, above_max=0, min=1, max=20
DaysLateLast30: below_min=0, above_max=0, min=0, max=6
SpecialProjectsCount: below_min=0, above_max=0, min=0, max=8
PerfScoreID: below_min=0, above_max=0, min=1, max=4
EmpStatusID: below_min=0, above_max=0, min=1, max=5


## Employee Numeric Range Validation

Key numeric fields in the Employee dataset were checked against expected non-negative or coded-value ranges.

| Field | Observed Minimum | Observed Maximum | Range Violations | Result |
|---|---:|---:|---:|---|
| `Salary` | 45,046 | 250,000 | 0 | Pass |
| `Absences` | 1 | 20 | 0 | Pass |
| `DaysLateLast30` | 0 | 6 | 0 | Pass |
| `SpecialProjectsCount` | 0 | 8 | 0 | Pass |
| `PerfScoreID` | 1 | 4 | 0 | Pass |
| `EmpStatusID` | 1 | 5 | 0 | Pass |

No range violations were identified. Salary is retained in the raw source for lineage but will be excluded from the MVP analytical layer because compensation is outside the approved project scope.

In [31]:
print("=" * 60)
print("DATE ORDER VALIDATION")
print("=" * 60)

# Employee: termination should not occur before hire
employee = datasets["Employee"]

hire = pd.to_datetime(employee["DateofHire"], errors="coerce")
termination = pd.to_datetime(employee["DateofTermination"], errors="coerce")

invalid_termination = (
    termination.notna()
    & hire.notna()
    & (termination < hire)
)

print(
    "Employee - termination before hire:",
    invalid_termination.sum()
)


# Learning: completion should not occur before assignment
learning = datasets["Learning Management"]

assigned = pd.to_datetime(
    learning["assigned_date"],
    errors="coerce"
)

completed = pd.to_datetime(
    learning["completed_date"],
    errors="coerce"
)

invalid_learning = (
    completed.notna()
    & assigned.notna()
    & (completed < assigned)
)

print(
    "Learning - completion before assignment:",
    invalid_learning.sum()
)


# Onboarding: completion before due date
onboarding = datasets["Onboarding Progress"]

due = pd.to_datetime(
    onboarding["due_date"],
    errors="coerce"
)

completed = pd.to_datetime(
    onboarding["completed_date"],
    errors="coerce"
)

early_completion = (
    completed.notna()
    & due.notna()
    & (completed < due)
)

print(
    "Onboarding - completed before due date:",
    early_completion.sum()
)

DATE ORDER VALIDATION
Employee - termination before hire: 0
Learning - completion before assignment: 0
Onboarding - completed before due date: 1788


## Date-Order Validation

Business date relationships were checked across the integrated datasets.

| Validation Rule | Result | Interpretation |
|---|---:|---|
| Employee termination date before hire date | 0 | Pass |
| Learning completion date before assignment date | 0 | Pass |
| Onboarding completion date before due date | 1,788 | Valid early completions |

No logically invalid employee or learning date sequences were identified.

The 1,788 onboarding records completed before their assigned due date are not treated as errors. They represent tasks completed ahead of schedule and may provide a useful operational signal for onboarding efficiency analysis.

In [32]:
profiling_summary = pd.DataFrame({
    "Dataset": [
        "Employee",
        "Onboarding Progress",
        "Internal Tool Usage",
        "Support Tickets",
        "Learning Management"
    ],
    "Rows": [
        len(datasets["Employee"]),
        len(datasets["Onboarding Progress"]),
        len(datasets["Internal Tool Usage"]),
        len(datasets["Support Tickets"]),
        len(datasets["Learning Management"])
    ],
    "Columns": [
        datasets["Employee"].shape[1],
        datasets["Onboarding Progress"].shape[1],
        datasets["Internal Tool Usage"].shape[1],
        datasets["Support Tickets"].shape[1],
        datasets["Learning Management"].shape[1]
    ],
    "Primary Key": [
        "EmpID",
        "onboarding_task_id",
        "usage_id",
        "ticket_id",
        "learning_record_id"
    ],
    "Duplicate IDs": [
        0,
        0,
        0,
        0,
        0
    ],
    "Null IDs": [
        0,
        0,
        0,
        0,
        0
    ],
    "Referential Integrity": [
        "N/A",
        "Pass",
        "Pass",
        "Pass",
        "Pass"
    ]
})

profiling_summary

,Dataset,Rows,Columns,Primary Key,Duplicate IDs,Null IDs,Referential Integrity
0,Employee,311,36,EmpID,0,0,N/A
1,Onboarding Progress,4222,9,onboarding_task_id,0,0,Pass
2,Internal Tool Usage,13995,9,usage_id,0,0,Pass
3,Support Tickets,1048,8,ticket_id,0,0,Pass
4,Learning Management,3192,9,learning_record_id,0,0,Pass


## Final Profiling Summary

| Dataset | Rows | Columns | Primary Key | Duplicate IDs | Null IDs | Referential Integrity |
|---|---:|---:|---|---:|---:|---|
| Employee | 311 | 36 | `EmpID` | 0 | 0 | N/A |
| Onboarding Progress | 4,222 | 9 | `onboarding_task_id` | 0 | 0 | Pass |
| Internal Tool Usage | 13,995 | 9 | `usage_id` | 0 | 0 | Pass |
| Support Tickets | 1,048 | 8 | `ticket_id` | 0 | 0 | Pass |
| Learning Management | 3,192 | 9 | `learning_record_id` | 0 | 0 | Pass |